src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs

```diff
@@ -0,0 +1,272 @@
+
+            public override bool MoveNext()
+            {
+                int state = _state;
+
+            Initialized:
+                if (state > 1)
+                {
+                    List<TSource>? buffer = _buffer;
+                    Debug.Assert(buffer is not null);
+
+                    int i = state - 2;
+                    if (i < buffer.Count)
+                    {
+                        _current = buffer[i];
+                        _state++;
+                        return true;
+                    }
+                }
+                else if (state == 1)
+                {
+                    List<TSource>? buffer = SampleToList(_source, _takeCount);
+                    if (buffer is not null)
+                    {
+                        _buffer = buffer;
+                        _state = state = 2;
+                        goto Initialized;
+                    }
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public override void Dispose()
+            {
+                _buffer = null;
+                base.Dispose();
+            }
+
+            /// <summary>Uses reservoir sampling to randomly select <paramref name="takeCount"/> elements from <paramref name="source"/>.</summary>
+            private static List<TSource>? SampleToList(IEnumerable<TSource> source, int takeCount)
+            {
+                List<TSource>? reservoir = null;
+
+                if (source is IList<TSource> list)
+                {
+                    int listCount = list.Count;
+                    Debug.Assert(listCount > takeCount, "Known listCount <= takeCount should have been handled by Iterator.Take override");
+
+                    reservoir = new(takeCount);
+                    for (int i = 0; i < listCount; i++)
+                    {
+                        if (i < takeCount)
+                        {
+                            // Fill the reservoir with the first takeCount elements from the source.
+                            reservoir.Add(list[i]);
+                        }
+                        else
+                        {
+                            // For each subsequent element in the source, randomly replace an element in the
+                            // reservoir with a decreasing probability.
+                            int r = Random.Shared.Next(i + 1);
+                            if (r < takeCount)
+                            {
+                                reservoir[r] = list[i];
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    using IEnumerator<TSource> e = source.GetEnumerator();
+                    if (e.MoveNext())
+                    {
+                        reservoir = [e.Current];
```

Comment:
Shouldn't the reservoir also be presized to `takeCount` here?

Reply:
We don't know how many elements will actually be in the source. It could be much less than the takeCount.